### PR TITLE
Improve unwind support on Windows x64.

### PR DESCRIPTION
**Pull request by lateralusX**
This commit improves the unwind support on Windows x64 by implementing a number of missing features. It also adds support for out of proc function table callback module on older OS versions and switch to different function table methods (growable function tables) on Win8/Win2012Server and later OS versions.

Commit aligns more to the windows x64 prolog/epilog ABI, but since mono uses negative offsets from frame pointer while x64 windows uses positive, it is not possible to do a 100% "correct" representation in the prolog, but if we don’t give the OS unwinder the full unwind metadata in the frame pointer use case,
we can trick unwinder to do the right thing presenting correct callstacks in debuggers.

Commit includes representation of windows unwind info in mono's unwind data using extensions to better integrate windows specific unwind info in areas currently not supported (like trampolines). It will also be easier to implement unwind info in full AOT objects in the future, if windows specific unwind data is included in mono’s unwind data structures.

On Win8/Win2012Server and later OS versions, all function table data is registered using growable function tables giving full support for callstacks on live debug targets and crash dumps. On previous OS versions function table callbacks are used together with an out of proc module. This works for live debugging in WinDBG and Visual Studio (if right permissions are given to the debuggers to load out of proc unwind module).

Crash dumps works as expected in WinDBG but currently not in Visual Studio for older OS versions using callback support. If that use case needs to be supported it needs to be solved in a separate PR, but since we support crash dumps in Visual Studio on later OS versions and in WinDBG on earlier, it is still possible to analyze crash dumps by picking the right tool.

Debuggers needs some additional information to work with function table callbacks out of proc modules (NOTE, this is NOT needed on Win8/Win2012Server and later).

WinDBG:

To just do live debugging, it is possible to run the following command in WinDBG,

settings set EngineInitialization.VerifyFunctionTableCallbacks=false

To do crash dump analysis the following registry key needs to be set,

HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\KnownFunctionTableDlls

Add the full path to mono-2.0-dac-sgen.dll matching the mono-2.0-sgen.dll as a 32-bit REG_DWORD.

NOTE, adding the registry key removes the need to set the setting.

Visual Studio:

To do live debugging, set the same registry key as WinDBG use case.

### CODE REVIEW

**Comment by kumpera** [comment]
@vargaz can you review the unwinding changes?

File: mono/mini/exceptions-amd64.c [hunk 1]
```diff
@@ -273,7 +274,12 @@ mono_arch_get_call_filter (MonoTrampInfo **info, gboolean aot)
 		if (AMD64_IS_CALLEE_SAVED_REG (i))
 			amd64_pop_reg (code, i);
 
+#if TARGET_WIN32
+	amd64_lea_membase (code, AMD64_RSP, AMD64_RBP, 0);
+	amd64_pop_reg (code, AMD64_RBP);
+#else
```
Comment (vargaz) [inline mono/mini/exceptions-amd64.c]: Isn't this the same ?

File: mono/mini/exceptions-amd64.c [hunk 2]
```diff
@@ -455,6 +466,7 @@ get_throw_trampoline (MonoTrampInfo **info, gboolean rethrow, gboolean corlib, g
 	mono_arch_flush_icache (start, code - start);
 
 	g_assert ((code - start) < kMaxCodeSize);
+	g_assert_checked (mono_arch_unwindinfo_validate_size (unwind_ops, MONO_MAX_TRAMPOLINE_UNWINDINFO_SIZE));
```
Comment (vargaz) [inline mono/mini/exceptions-amd64.c]: This could be a normal assert, _checked  () asserts are mainly useful in perf sensitive code.

File: mono/mini/exceptions-amd64.c [hunk 3]
```diff
@@ -885,260 +897,878 @@ mono_arch_exceptions_init (void)
...
+#define MONO_DAC_MODULE L"mono-2.0-dac-sgen.dll"
+#else
+#define MONO_DAC_MODULE L"mono-2.0-sgen.dll"
+#endif
+
+#define MONO_DAC_MODULE_MAX_PATH 1024
+
+static void
+mono_arch_unwindinfo_init_table_no_lock (void)
 {
```
Comment (vargaz) [inline mono/mini/exceptions-amd64.c]: Static functions don't need a long prefix.

File: mono/mini/exceptions-amd64.c [hunk 4]
```diff
@@ -885,260 +897,878 @@ mono_arch_exceptions_init (void)
 	}
 }
 
-#if defined(TARGET_WIN32) && !defined(DISABLE_JIT)
-
```
Comment (vargaz) [inline mono/mini/exceptions-amd64.c]: Why is this removed ?

File: mono/mini/exceptions-amd64.c (hunk 1 above)
Comment (lateralusX) [inline mono/mini/exceptions-amd64.c]: The windows x64 ABI only handle a fixed set of epilog patterns that must be followed, so even if the result will be the same as using amd64_leave, it is not an instruction sequence that is approved to be part of an x64 Windows compliant epilog. Short story:

No FP:
add RSP, fixed-allocation-size  
pop's matching prolog pushes
ret  

FP:
lea      RSP, fixed-allocation-size – 128[RBP]  
pop's matching prolog pushes
ret

These are the only legal forms for an epilog. It must consist of either an add RSP,constant or lea RSP,constant[FPReg], followed by a series of zero or more 8-byte register pops and a return or a jmp

More details:

https://msdn.microsoft.com/en-us/library/tawsa7cb.aspx

File: mono/mini/exceptions-amd64.c (hunk 2 above)
Comment (lateralusX) [inline mono/mini/exceptions-amd64.c]: I put it into a checked assert since the mono_arch_unwindinfo_validate_size does some additional work to walk through the list of unwind ops and classify them (the same that is done later when installing it into the unwind tables) accordingly to calculate size needed to serialize them. This work is only needed to validate the pre allocated buffer that in turn builds on a size assumption done in code. It is not depending on any input data, so as long as  the code doesn't change (adding new or change unwind ops used by windows) it will always be true. The performance impact will be in direct relationship with how many trampolines that gets created, so if this is not that frequently called in relation to everything else, we could make it a normal g_assert, but then we will do the calculation twice in release builds where it most likely should be true 100% of the times. I think this is something you normally would like to get noticed during development and fixed and then don't paid the additional cost in release builds. If the checked build assert (something we will run on windows debug builds during development) is a problem, I can switch to a normal g_assert. Thoughts?

File: mono/mini/exceptions-amd64.c (hunk 3 above)
Comment (lateralusX) [inline mono/mini/exceptions-amd64.c]: Used the same naming pattern for consistency, no problem to shorten the static function names. I will then drop mono_arch_unwindinfo prefix for these methods ending up with names like, init_table_no_lock, find_range_in_table_no_lock etc. OK?

File: mono/mini/exceptions-amd64.c (hunk 2 above)
Comment (vargaz) [inline mono/mini/exceptions-amd64.c]: Ok, didn't notice that its costly.

File: mono/mini/exceptions-amd64.c (hunk 3 above)
Comment (vargaz) [inline mono/mini/exceptions-amd64.c]: Ok.

File: mono/mini/exceptions-amd64.c (hunk 1 above)
Comment (vargaz) [inline mono/mini/exceptions-amd64.c]: Ok.
